Blob should report its underlying memory cost to garbage collection

cdumez opened the pull request:
#### c51b8f774888924b36573b76dd14300d216b1166
<pre>
Blob should report its underlying memory cost to garbage collection
<a href="https://bugs.webkit.org/show_bug.cgi?id=244504">https://bugs.webkit.org/show_bug.cgi?id=244504</a>
&lt;rdar://99288771&gt;

Reviewed by Youenn Fablet.

* Source/WebCore/Modules/fetch/FetchBody.cpp:
(WebCore::FetchBody::fromFormData):
* Source/WebCore/Modules/websockets/WorkerThreadableWebSocketChannel.cpp:
(WebCore::WorkerThreadableWebSocketChannel::Bridge::send):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::computeMemoryCost):
(WebCore::Blob::Blob):
* Source/WebCore/fileapi/Blob.h:
(WebCore::Blob::deserialize):
(WebCore::Blob::memoryCost const):
* Source/WebCore/fileapi/Blob.idl:
* Source/WebCore/fileapi/File.cpp:
(WebCore::File::File):

Canonical link: <a href="https://commits.webkit.org/254026@main">https://commits.webkit.org/254026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d710a80f55abfa7b8039cfdc9ae9da36cb8cac28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18504 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96968 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/151044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30226 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79875 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91712 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93391 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24415 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74513 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24399 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79373 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67246 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27910 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27884 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14371 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2828 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29571 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37283 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29467 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33647 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -183,6 +204,7 @@ Blob::Blob(DeserializationContructor, ScriptExecutionContext* context, const URL
 Blob::Blob(ScriptExecutionContext* context, const URL& srcURL, long long start, long long end, const String& type)
     : ActiveDOMObject(context)
     , m_type(normalizedContentType(type))
+    , m_memoryCost(end - start)
```
This is probably incorrect in case end is for instance -1, or if slicing a File.
It might be better to pass a memoryCost parameter and compute its value in Blob::slice.

youennf reviewed (approved):
LGTM overall.
I guess the most accurate approach would be for NetworkProcess to report precisely what a context or a page is actually using. But this might be overkill.

cdumez committed:
Blob should report its underlying memory cost to garbage collection
https://bugs.webkit.org/show_bug.cgi?id=244504
<rdar://99288771>

Reviewed by Youenn Fablet.

* Source/WebCore/Modules/fetch/FetchBody.cpp:
(WebCore::FetchBody::fromFormData):
* Source/WebCore/Modules/websockets/WorkerThreadableWebSocketChannel.cpp:
(WebCore::WorkerThreadableWebSocketChannel::Bridge::send):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::computeMemoryCost):
(WebCore::Blob::Blob):
* Source/WebCore/fileapi/Blob.h:
(WebCore::Blob::deserialize):
(WebCore::Blob::memoryCost const):
* Source/WebCore/fileapi/Blob.idl:
* Source/WebCore/fileapi/File.cpp:
(WebCore::File::File):

Canonical link: https://commits.webkit.org/254026@main
